src/software/new_geom/cubic_bezier_spline2d.h

```diff
@@ -0,0 +1,124 @@
+     *
+     *
+     * @param start_point The first point on the spline
+     * @param start_vector The tangent vector at the start of the spline
+     * @param end_point The last point on the spline
+     * @param end_vector The tangent vector at the end of the spline.
+     *                   NOTE: this points "back towards" the start of the spline, instead
+     *                         of "off the end" of the spline. ie. this vector is
+     *                         "symmetric" with `start_vector`
+     * @param intermediate_knots The intermediate points between `start_point` and
+     *                                    `end_point` that this spline will interpolate
+     */
+    explicit CubicBezierSpline2d(const Point& start_point, const Vector& start_vector,
+                                 const Point& end_point, const Vector& end_vector,
+                                 const std::vector<Point>& intermediate_knots);
+
+    const Point getValueAt(double t) const override;
+    const std::vector<Point> getKnots() const override;
+    size_t getNumKnots() const override;
+    std::vector<double> getKnotParametrizationValues() const override;
+    const Point getStartPoint() const override;
+    const Point getEndPoint() const override;
+    const std::vector<SplineSegment2d> getSplineSegments() const override;
+
+    /**
+     * Get the control points that are the implicit representation of the bezier
+     * curves that form this spline.
+     *
+     * @return The control points that represent the the bezier curves that form this
+     *         spline.
+     */
+    const std::vector<Point>& getControlPoints() const;
+
+    /**
+     * Get the number of segments on this spline.
+     *
+     * @return The number of segments on this spline.
+     */
+    size_t getNumSegments() const;
+
+
+   private:
+    /**
+     * Get the segment of this spline at the given index
+     *
+     * @param index The index of the segment on this spline. Must be in
+     *              [0, getNumSegments()-1]
+     *
+     * @return The segment of this spline at the given index
+     */
+    BezierCurve2d getSegmentAtIndex(size_t index) const;
+
+    /**
+     * Compute the control points that represent the bezier curves that satisfy the
+     * constraints of this spline
+     *
+     * @param start_point The first point on the spline
+     * @param start_vector The tangent vector at the start of the spline
+     * @param end_point The last point on the spline
+     * @param end_vector The tangent vector at the end of the spline.
+     *                   NOTE: this points "back towards" the start of the spline, instead
+     *                         of "off the end" of the spline. ie. this vector is
+     *                         "symmetric" with `start_vector`
+     * @param intermediate_knots The intermediate points between `start_point` and
+     *                                    `end_point` that this spline will interpolate
```

Comment:
Interpolate *between*?

Reply:
This is unnecessarily mathy, changed to `pass through`